File: sdk/go/direct_test.go
package validator

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"testing"
)

type directTransport func(*http.Request) (*http.Response, error)

func (f directTransport) RoundTrip(request *http.Request) (*http.Response, error) { return f(request) }

func TestDirectClient(t *testing.T) {
	var calls int
	transport := directTransport(func(r *http.Request) (*http.Response, error) {
		calls++
		if r.URL.String() != "https://api.typesafe.ai/v1/systemone" || r.Header.Get("Authorization") != "Bearer own-key" {
			t.Errorf("wrong endpoint or bearer header")
		}
		var body struct {
			Model     string `json:"model"`
			State     string `json:"state"`
			Questions map[string]struct {
				Type         string `json:"type"`
				Instructions string `json:"instructions"`
			} `json:"questions"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Error(err)
		}
		if body.Model != "jev-latest" || body.State == "" || body.Questions["result"].Type != "choice" {
			t.Errorf("wrong Jev payload: %+v", body)
		}
		choice, confidence := "true", "0.93"
		if body.State == "No" {
			choice, confidence = "false", "0.98"
		}
		if body.State == "Maybe" {
			confidence = "0.63"
		}
		response := `{"answers":{"result":{"type":"choice","choice":"` + choice + `","confidence":` + confidence + `}}}`
		return &http.Response{StatusCode: 200, Body: io.NopCloser(strings.NewReader(response)), Header: make(http.Header)}, nil
	})
	client, err := NewDirectClientWithOptions("own-key", DirectOptions{HTTPClient: &http.Client{Transport: transport}})
	if err != nil {
		t.Fatal(err)
	}
	result, err := client.Name(context.Background(), "Jorge Eduardo")
	if err != nil || result.Valid == nil || !*result.Valid || result.Status != StatusValid {
		t.Fatalf("unexpected result: %+v %v", result, err)
	}
	negative, err := client.Check(context.Background(), "No", "¿Compra?")
	if err != nil || negative.Valid == nil || *negative.Valid || negative.Status != StatusInvalid {
		t.Fatalf("unexpected negative result: %+v %v", negative, err)
	}
	uncertain, err := client.Check(context.Background(), "Maybe", "Question")
	if err != nil || uncertain.Valid != nil || uncertain.Status != StatusUncertain {
		t.Fatalf("unexpected uncertain result: %+v %v", uncertain, err)
	}
	batch, err := client.ValidateBatch(context.Background(), []BatchItem{
		{ID: "a", Rule: "person_name", Value: "Jorge"},
		{ID: "b", Rule: "missing", Value: "No"},
		{ID: "c", Rule: "person_name", Value: " "},
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(batch) != 3 || batch[0].ID != "a" || batch[1].Error.Code != "unknown_rule" || batch[2].Error.Code != "invalid_request" {
		t.Fatalf("unexpected batch: %+v", batch)
	}
	if calls != 4 {
		t.Fatalf("expected 4 Jev calls, got %d", calls)
	}
}

func TestDirectInvalidResponseAndConfig(t *testing.T) {
	if _, err := NewDirectClient(""); err == nil {
		t.Fatal("missing key accepted")
	}
	if _, err := NewDirectClientWithOptions("key", DirectOptions{BaseURL: "http://example.com"}); err == nil {
		t.Fatal("insecure endpoint accepted")
	}
	transport := directTransport(func(*http.Request) (*http.Response, error) {
		return &http.Response{StatusCode: 200, Body: io.NopCloser(strings.NewReader(`{}`)), Header: make(http.Header)}, nil
	})
	client, err := NewDirectClientWithOptions("key", DirectOptions{HTTPClient: &http.Client{Transport: transport}})
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.Check(context.Background(), "text", "question")
	if directErr, ok := err.(*DirectError); !ok || directErr.Code != "invalid_response" {
		t.Fatalf("wrong error: %v", err)
	}
}

File: sdk/typescript/test/direct.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { DirectValidator, Validator } from "../dist/index.js";

function mock(choice = "true", confidence = 0.95) {
  const calls = [];
  const fetcher = async (url, options) => {
    calls.push({ url, options });
    return new Response(JSON.stringify({ answers: { result: { type: "choice", choice, confidence } } }), {
      status: 200, headers: { "Content-Type": "application/json" },
    });
  };
  return { client: new DirectValidator({ jevApiKey: "test-only", fetcher }), calls };
}

test("direct client calls Jev with own key and returns statuses", async () => {
  const { client, calls } = mock();
  assert.deepEqual(await client.name(" Jorge Eduardo "), { valid: true, confidence: 0.95, status: "valid" });
  assert.equal(calls[0].url, "https://api.typesafe.ai/v1/systemone");
  assert.equal(calls[0].options.headers.Authorization, "Bearer test-only");
  const body = JSON.parse(calls[0].options.body);
  assert.equal(body.model, "jev-latest");
  assert.equal(body.state, "Jorge Eduardo");
  assert.equal(body.questions.result.type, "choice");
  assert.match(body.questions.result.instructions, /nombre de una persona/);
  assert.equal((await mock("false", 0.98).client.check("No", "¿Compra?")).status, "invalid");
  assert.deepEqual(await mock("true", 0.63).client.check("texto", "pregunta"), { valid: null, confidence: 0.63, status: "uncertain" });
});

test("direct batch keeps order and reports item errors", async () => {
  const { client, calls } = mock();
  const result = await client.validateBatch([
    { id: "a", rule: "person_name", value: "Jorge" },
    { id: "b", rule: "missing", value: "No" },
    { id: "c", rule: "person_name", value: " " },
  ]);
  assert.equal(result.length, 3);
  assert.deepEqual(result.map((item) => item.id), ["a", "b", "c"]);
  assert.equal(result[1].error.code, "unknown_rule");
  assert.equal(result[2].error.code, "invalid_request");
  assert.equal(calls.length, 1);
});

test("direct client sends versioned field-rule questions", async () => {
  const { client, calls } = mock();
  await client.validate("product_description", "Taladro de 18 V");
  await client.validate("address", "Av. Bolívar 123");
  assert.match(JSON.parse(calls[0].options.body).questions.result.instructions, /característica concreta/);
  assert.match(JSON.parse(calls[1].options.body).questions.result.instructions, /dirección física/);
});

test("direct client rejects missing keys and invalid provider responses", async () => {
  assert.throws(() => new DirectValidator({ jevApiKey: "" }), /jevApiKey/);
  assert.throws(() => new DirectValidator({ jevApiKey: "key", baseURL: "http://example.com" }), /api.typesafe.ai/);
  assert.throws(() => new DirectValidator({ jevApiKey: "key", baseURL: "https://validator.trialsur.cloud" }), /api.typesafe.ai/);
  const client = new DirectValidator({ jevApiKey: "key", fetcher: async () => new Response("{}") });
  await assert.rejects(() => client.check("value", "question"), { code: "invalid_response" });
  const fail = new DirectValidator({ jevApiKey: "key", fetcher: async () => new Response("error", { status: 401 }) });
  await assert.rejects(() => fail.check("value", "question"), { code: "provider_error", statusCode: 401 });
  const legacy = new Validator({ apiKey: "local-dev", fetcher: async () => new Response(JSON.stringify({ valid: true, confidence: 1, status: "valid" })) });
  assert.equal((await legacy.name("Jorge")).valid, true);
});
